### PR TITLE
Add Dropout ONNX OP and other refactoring

### PR DESCRIPTION
**Pull request by antimora**
Add full support for Dropout OP since we want to support importing trainable models. Instead of removing Dropout like other ONNX runtimes, I opted to support the Dropout OP. [ONNX Runtime by Microsoft](https://onnxruntime.ai/) is pushing for the ability to train ONNX models. PyTorch can export models in trainable mode.

### Checklist

- [x] Confirm that `run-checks.sh` has been executed.

### Changes

1. Add Dropout OP.
2. Support Constant lifting: some inputs use a Constant node to initialize data. For consistency, the Constants are moved to states. This process is called Constant lifting.
3. Add Boolean scalar type.
4. Add a phantom field for all cases, not only the empty struct, because there are instances when fields may not contain `<B>` generics.

### Testing

Tested a generated ONNX file (checked in) and a SqueezeNet ONNX file.

### CODE REVIEW

File: burn-import/src/onnx/from_onnx.rs [hunk 1]
```diff
@@ -428,10 +431,87 @@ fn move_inputs_to_state(nodes: &mut Vec<Node>, initializer: &[TensorProto]) {
...
+/// and consistently use the same interface for all nodes (constant inputs and inputs with initializers are
+/// treated the same way).
+///
+/// # Arguments
+///
+/// * `nodes` - A mutable reference to a vector of nodes
+///
+/// # Panics
+///
+/// Panics if the node's output is not a constant.
```
Comment (nathanielsimard) [inline burn-import/src/onnx/from_onnx.rs:450]: I think an example would help here to understand what the function actually do (how does the graph look before and after).

**Review by nathanielsimard — COMMENTED** [review]
I have some questions, but overall looks good!

File: burn-import/src/burn/node/base.rs [hunk 2]
```diff
@@ -204,13 +207,14 @@ pub(crate) mod tests {
 
             #[derive(Module, Debug)]
             pub struct Model<B: Backend> {
-                _phantom: core::marker::PhantomData<B>,
+                phantom: core::marker::PhantomData<B>,
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/burn/node/base.rs:210]: I thought as well that it will come out as unused variable but since it's a field compiler does not mind. I used `phantom` for aesthetic reasons and I have seen this is how used. For example, the [examples in PhantomData](https://doc.rust-lang.org/beta/core/marker/struct.PhantomData.html) use `phantom` field name instead of `_phantom`.  So I didn't want to stand out =)

The reason that I am using `#[allow(unused_variables)]` for record in `with_init`. Unless I peak into the body's code, I can't decide if we should use `_record` or `record`. For simplicity and avoid error prone implementation, I opted to use `#[allow(unused_variables)]`. Of course doing so blankets the coverage, and in fact there is a variable unused in the body, we will not know. What do you think about this aspect? Should we check body's code if there is a reference to `record` argument? BTW, Dropout needs a field but does not need record, as you know.

[the remark this answers is not in the document]
Comment (nathanielsimard) [reply burn-import/src/burn/node/base.rs:210]: No let's keep `#[allow(unused_variable)]` for simplicity, we can revisit later if it becomes a problem.

File: burn-import/src/onnx/from_onnx.rs [hunk 3]
```diff
@@ -428,10 +431,87 @@ fn move_inputs_to_state(nodes: &mut Vec<Node>, initializer: &[TensorProto]) {
...
+fn lift_constants(nodes: &mut Vec<Node>) {
+    log::info!("Lifting constants into the states");
+
+    // create a set to hold the node types to process
+    let node_types_to_process: HashSet<NodeType> =
+        [NodeType::Dropout, NodeType::Conv1d, NodeType::Conv2d]
+            .into_iter()
+            .collect();
```
[the remark this answers is not in the document]
Comment (nathanielsimard) [reply burn-import/src/onnx/from_onnx.rs:467]: Thanks for the explaination 🙏

**Review by nathanielsimard — APPROVED** [review]

**Comment by nathanielsimard** [comment]
@antimora Do you know why no check has been run? Maybe if you merge main in your branch and push it will trigger the CI?

**Comment by antimora** [comment]
> @antimora Do you know why no check has been run? Maybe if you merge main in your branch and push, it will trigger the CI?

No idea. I was wondering too. I thought I needed some permission.
